File: back/app/services/alerts.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.regional_dashboard import Client, RegionalAlert
from app.schemas.alerts import SalesAlertResponse

_RISK_TO_FRONTEND = {"high": "high", "medium": "medium", "low": "low"}


async def get_alerts(session: AsyncSession) -> list[SalesAlertResponse]:
    result = await session.execute(
        select(RegionalAlert, Client)
        .join(Client, RegionalAlert.client_id == Client.id)
        .order_by(RegionalAlert.created_at.desc())
    )
    rows = result.all()

    alerts: list[SalesAlertResponse] = []
    for alert, client in rows:
        alerts.append(
            SalesAlertResponse(
                id=str(alert.id),
                clientName=client.name,
                riskLevel=_RISK_TO_FRONTEND.get(alert.risk_level, "medium"),
                churnProbability=alert.churn_probability,
                purchasePropensity=alert.purchase_propensity,
                customerValue=_RISK_TO_FRONTEND.get(client.customer_value, "medium"),
                explanation=alert.explanation or "",
                churnType=alert.churn_type or "Total",
                status=alert.status,
                interactions=[],
                events=[],
                alertContextJson=alert.alert_context_json,
                predictedNextPurchase=alert.predicted_next_purchase,
                lastOrderDate=alert.last_order_date,
            )
        )
    return alerts

File: front/src/api/alerts.ts
import { apiRequest } from "@/api/client";
import type { SalesAlert } from "@/types/alerts";

type AlertApiRow = {
  id: string;
  clientName: string;
  riskLevel: "low" | "medium" | "high";
  churnProbability: number;
  purchasePropensity: number;
  customerValue: "low" | "medium" | "high";
  explanation: string;
  churnType: string;
  status: string;
  interactions: never[];
  events: never[];
  alertContextJson?: string | null;
  predictedNextPurchase?: string | null;
  lastOrderDate?: string | null;
};

export async function fetchAlerts(agentId?: number): Promise<SalesAlert[]> {
  const qs = agentId !== undefined ? `?agent_id=${agentId}` : "";
  const rows = await apiRequest<AlertApiRow[]>(`/alerts${qs}`);
  return rows.map((r) => ({
    id: r.id,
    clientName: r.clientName,
    riskLevel: r.riskLevel,
    churnProbability: r.churnProbability,
    purchasePropensity: r.purchasePropensity,
    customerValue: r.customerValue,
    explanation: r.explanation,
    churnType: r.churnType,
    status: (r.status as SalesAlert["status"]) ?? "pending",
    interactions: [],
    events: [],
    alertContextJson: r.alertContextJson ?? undefined,
    predictedNextPurchase: r.predictedNextPurchase ?? undefined,
    lastOrderDate: r.lastOrderDate ?? undefined,
  }));
}
